### ISSUE TITLE
rstanarm with random effects: "Error: couldn't evaluate grouping factor ... within model frame"

### ISSUE THREAD

**Issue by DominiqueMakowski** [rvlenth/emmeans#204]
Continuing my analyses, I stumbled across another issue which I'm not sure is a bug or myself doing something wrong.

I looked for help on the internet but only found this error message directly related to lme4 and some scooping variables problems (example [here](https://stackoverflow.com/questions/36650695/grouping-error-with-lmer) but doesn't look like it's related to this error), so decided to post it here nonetheless.

Essentially, I didn't manage to make emmeans work with lme4/gamm4 (which I think is built on top of lme4) random effects:

``` r
library(rstanarm)

model <- rstanarm::stan_gamm4(Petal.Length ~ s(Sepal.Width), random=~(1|Species), data=iris, chains=2, iter=200, refresh=0)

emmeans::ref_grid(model)
#> Error: couldn't evaluate grouping factor Species within model frame: try adding grouping factor to data frame explicitly if possible
emmeans::ref_grid(model, at=list("Sepal.Width"=c(2,3,4), "Species"=c("setosa")))
#> Error: couldn't evaluate grouping factor Species within model frame: try adding grouping factor to data frame explicitly if possible


model <- rstanarm::stan_lmer(Petal.Length ~ Sepal.Width + (1|Species), data=iris, chains=2, iter=200, refresh=0)

emmeans::ref_grid(model)
#> Error in eval(predvars, data, env): object 'Species' not found
emmeans::ref_grid(model, "Sepal.Width")
#> Error in eval(predvars, data, env): object 'Species' not found


# However, this works
model <- lme4::lmer(Petal.Length ~ Sepal.Width + (1|Species), data=iris)
emmeans::ref_grid(model)
#> 'emmGrid' object with variables:
#>     Sepal.Width = 3.0573
```

<sup>Created on 2020-05-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Thanks in advance!

**Comment by rvlenth** [comment]
Hmmm. The simple explanation is I didn't check the default for the `re.form` argument in `posterior_predict`. Given that, I thought it'd be a simple fix, and in fact I thought you could work around it simply by passing `re.form = ~0` which would get passed on via `...`. However, that exposed something really odd. Apparently, if there are random effects, then `vcov()` no longer works the same way:
```r
> fixef(model)
     (Intercept) s(Sepal.Width).1 s(Sepal.Width).2 s(Sepal.Width).3 s(Sepal.Width).4 
     3.511897689      0.008984523      0.002615105      0.021434277     -0.123299476 
s(Sepal.Width).5 s(Sepal.Width).6 s(Sepal.Width).7 s(Sepal.Width).8 s(Sepal.Width).9 
    -0.113552359      0.229375160     -0.301606259     -0.246623426      0.136190685 
> vcov(model)
            (Intercept)
(Intercept)   0.9814864
```
(note the spline parameters are excluded in `vcov()`) -- compared to the fixed-effects in #202 where the spline parameters are included in `vcov()`
```r
> coef(model2)
      (Intercept) Speciesversicolor  Speciesvirginica s(Petal.Length).1 
        6.4881892        -0.7295862        -1.2103498         0.3845727 
s(Petal.Length).2 s(Petal.Length).3 s(Petal.Length).4 s(Petal.Length).5 
       -0.9204837        -0.8178837        -0.4436711        -0.3480465 
s(Petal.Length).6 s(Petal.Length).7 s(Petal.Length).8 s(Petal.Length).9 
       -1.3247956        -0.3108035         0.1139369         1.2938412 
> vcov(model2)
                   (Intercept) Speciesversicolor Speciesvirginica s(Petal.Length).1
(Intercept)        0.099938313       -0.14864272     -0.152172495       -0.04977096
Speciesversicolor -0.148642722        0.22670148      0.223728329        0.07111805
Speciesvirginica  -0.152172495        0.22372833      0.236920295        0.07104810
s(Petal.Length).1 -0.049770963        0.07111805      0.071048098        1.27689667
s(Petal.Length).2  0.065322561       -0.10553318     -0.085279672       -0.30651586
s(Petal.Length).3  0.156244190       -0.24447277     -0.226549017       -0.25971461
s(Petal.Length).4  0.069656792       -0.10262026     -0.107493318        0.14204761
s(Petal.Length).5 -0.062349012        0.08759735      0.100853756       -0.09087337
s(Petal.Length).6  0.153912867       -0.25337354     -0.211032040       -0.21497889
s(Petal.Length).7 -0.028403510        0.04123317      0.046330845       -0.07606922
s(Petal.Length).8  0.005894863       -0.01022435     -0.004174682       -0.05994470
s(Petal.Length).9  0.512237602       -0.80014529     -0.740888606       -1.10260896
                  s(Petal.Length).2 s(Petal.Length).3 s(Petal.Length).4
(Intercept)              0.06532256        0.15624419        0.06965679

    ... etc ...
```
That noted, `fixef()` and `vcov()` are pretty unimportant here because it's the posterior samples that matter. So I just get my estimates from the sample. 
```r
> rg = ref_grid(model, at = list(Sepal.Width = 1:6))
> rg
'emmGrid' object with variables:
    Sepal.Width = 1, 2, 3, 4, 5, 6

> plot(rg)
```
![image](https://user-images.githubusercontent.com/26777519/83289780-262a3000-a1ab-11ea-9a2e-4f1ebc5ca0e8.png)

**Comment by rvlenth** [comment]
I think this is addressed, so am closing
